aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandler.java

```diff
@@ -17,42 +21,90 @@
-        final CallbackContext callbackContext,
-        final Logger logger) {
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final Logger logger) {
 
-        final ResourceModel model = request.getDesiredResourceState();
         final String awsAccountId = request.getAwsAccountId();
-        final String idempotencyToken = request.getClientRequestToken();
+        final ResourceModel model = request.getDesiredResourceState();
+        final String pgName = model.getProfilingGroupName();
 
+        CreateProfilingGroupRequest createProfilingGroupRequest = CreateProfilingGroupRequest.builder()
+                .profilingGroupName(pgName)
+                .clientToken(request.getClientRequestToken())
+                .build();
         try {
-            CreateProfilingGroupRequest createProfilingGroupRequest = CreateProfilingGroupRequest.builder()
-                    .profilingGroupName(model.getProfilingGroupName())
-                    .clientToken(idempotencyToken)
-                    .build();
-
             proxy.injectCredentialsAndInvokeV2(createProfilingGroupRequest, profilerClient::createProfilingGroup);
+        } catch (CodeGuruProfilerException e) {
+            wrapException(e);
+        }
+        logger.log(format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, pgName, awsAccountId));
+
+        Optional<List<String>> principals = principalsForAgentPermissionsFrom(model);
+        if (principals.isPresent()) {
+            putAgentPermissions(proxy, pgName, principals.get());
+            logger.log(format("%s [%s] for accountId [%s] has been successfully updated with agent permissions!",
+                    ResourceModel.TYPE_NAME, pgName, awsAccountId));
+        }
+
+        return ProgressEvent.defaultSuccessHandler(model);
+    }
 
-            logger.log(String.format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, model.getProfilingGroupName(), awsAccountId));
+    private void putAgentPermissions(final AmazonWebServicesClientProxy proxy, final String pgName, final List<String> principals) {
+        PutPermissionRequest putPermissionRequest = PutPermissionRequest.builder()
+                .profilingGroupName(pgName)
+                .actionGroup(ActionGroup.AGENT_PERMISSIONS)
+                .principals(principals)
+                .build();
+        try {
+            proxy.injectCredentialsAndInvokeV2(putPermissionRequest, profilerClient::putPermission);
+        } catch (CodeGuruProfilerException e) {
+            try {
+                DeleteProfilingGroupRequest deletePgRequest = DeleteProfilingGroupRequest.builder().profilingGroupName(pgName).build();
+                proxy.injectCredentialsAndInvokeV2(deletePgRequest, profilerClient::deleteProfilingGroup);
```

Comment:
Suggestion: I think it maybe quite confusing to customer if this happens without having a clear error message from the Cloudformation. 

For example, if a client creating a PG with permission and the PG creation succeeded while the putPermission failed, we would attempt to delete the PG on our side. However, if that deletion failed, the customer will see `Failed to create PG` while the PG exists in their console; rerunning the Cloudformation may/ may not fail due to PG already exists.

I would suggest (I am not sure if that works or not) to include a better error message when we are wrapping our exception with CfnException describing what has happened and recommend customer what to do next. Either rerunning it or (worse) ask the customer to delete the PG manually. (<- This require you to do a quick test to check how it handles exception message and whether rerun works in this scenario :D)

Reply:
<img width="1259" alt="Screenshot 2020-05-22 at 20 42 15" src="https://user-images.githubusercontent.com/65159577/82695126-31420500-9c6d-11ea-81d3-b7f7c6a4062d.png">

Unfortunately, the CloudFormation error that customers see in the console is very minimalist, see attached screenshot while done when testing with PutPermissions and DeletePG to fail.

Reply:
Thanks for running the test for that :). Shame about it ><...